url-plugins/src/main/java/co/cask/hydrator/plugin/realtime/UrlFetchRealtimeSource.java

```diff
@@ -0,0 +1,108 @@
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.realtime.RealtimeSource;
+import co.cask.cdap.etl.api.realtime.SourceState;
+import co.cask.hydrator.plugin.realtime.config.UrlFetchRealtimeSourceConfig;
+import com.google.common.base.Joiner;
+import com.google.common.base.Strings;
+import com.google.common.io.ByteStreams;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Real Time Source to fetch data from a url.
+ */
+@Plugin(type = RealtimeSource.PLUGIN_TYPE)
+@Name("URLFetch")
+@Description("Fetch data from an external URL at a regular interval.")
+public class UrlFetchRealtimeSource extends RealtimeSource<StructuredRecord> {
+  private static final Logger LOG = LoggerFactory.getLogger(UrlFetchRealtimeSource.class);
+
+  private static final Schema DEFAULT_SCHEMA = Schema.recordOf(
+    "event",
+    Schema.Field.of("ts", Schema.of(Schema.Type.LONG)),
+    Schema.Field.of("url", Schema.of(Schema.Type.STRING)),
+    Schema.Field.of("headers", Schema.mapOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.STRING))),
+    Schema.Field.of("body", Schema.of(Schema.Type.STRING))
+  );
+
+  private final UrlFetchRealtimeSourceConfig config;
+
+  public UrlFetchRealtimeSource(UrlFetchRealtimeSourceConfig config) {
+    this.config = config;
+  }
+
+  @Nullable
+  @Override
+  public SourceState poll(Emitter<StructuredRecord> writer, SourceState currentState) throws Exception {
+    URL url = new URL(config.getUrl());
+    String response;
+    HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+    try {
+      if (connection.getResponseCode() == HttpURLConnection.HTTP_OK) {
+        response = Bytes.toString(ByteStreams.toByteArray(connection.getInputStream()));
+      } else if (connection.getResponseCode() == HttpURLConnection.HTTP_BAD_REQUEST) {
+        response = Bytes.toString(ByteStreams.toByteArray(connection.getErrorStream()));
+      } else {
+        throw new Exception("Invalid response code returned: " + connection.getResponseCode());
+      }
+    } finally {
+      connection.disconnect();
+    }
+
+    Map<String, List<String>> headers = connection.getHeaderFields();
+    Map<String, String> flattenedHeaders = new HashMap<>();
+    for (Map.Entry<String, List<String>> entry : headers.entrySet()) {
+      if (!Strings.isNullOrEmpty(entry.getKey())) {
+        // If multiple values for the same header exist, concatenate them
+        flattenedHeaders.put(entry.getKey(), Joiner.on(',').skipNulls().join(entry.getValue()));
+      }
+    }
+    writeDefaultRecords(writer, response, flattenedHeaders);
+    TimeUnit.SECONDS.sleep(config.getIntervalInSeconds());
```

Comment:
the sleep will not occur if there was an exception before this point.  If you move this to the start of the method, it would guarantee the sleep always happens.